### PR TITLE
perf(gdn): use flashinfer sm100 prefill kernel for Qwen3.5 GDN

### PR DESCRIPTION
**Pull request by elwhyjay**
## What
Qwen3.5 GDN (gated delta net) prefill runs Triton FLA kernels. flashinfer 0.6.11 ships an sm100 GDN prefill kernel that is faster on Blackwell but unused. This adds it as a fast-path with a Triton fallback.
- ops/gated_delta_rule/flashinfer.py: kernel-package wrapper. Only the flashinfer path; never falls back, so it has no dependency on the runtime FLA code. is_supported() gates sm100 + head_dim==128 + bf16.
- hybrid_linear_attn.py: the runtime owns the branch (sm100 + head_dim==128 + bf16 + not need_h_track -> wrapper, else FLA). need_h_track needs intermediate states the flashinfer high-level API does not expose.
## Convention (vs FLA)
q/k L2-normalized by caller (sm100 kernel ignores use_qk_l2norm); g passed as alpha=exp(g) (kernel logs it internally); beta float32; recurrent state cast to float32 and transposed K<->V in/out; scale=head_dim**-0.5.
## Test / perf
test_gdn_flashinfer_fastpath: equal-head + GVA (ratio 1/2/4), zero/nonzero state, state dtype fp32/bf16, 25 passed. FLA uses atomic accumulation so its state is nondeterministic per-element (~0.45); the test asserts mean diff (<1e-3) as the correctness bar. Microbench (B200, Qwen3.5 GVA, T 2048-16384): ~2-7x.

| T | FLA (us) | fast-path (us) | speedup |
|---|---|---|---|
| 2048 | 1284 | 188 | 6.85x |
| 4096 | 910 | 210 | 4.33x |
| 8192 | 1198 | 375 | 3.20x |
| 16384 | 1505 | 700 | 2.15x |

### CODE REVIEW

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/gated_delta_rule/flashinfer.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        q3.contiguous(),
+        k3.contiguous(),
+        v3.contiguous(),
+        g=torch.exp(g2).float().contiguous(),
+        beta=beta2.float().contiguous(),
+        scale=scale,
+        # flashinfer requires fp32 state; runtime ssm dtype may be bf16.
+        initial_state=initial_state.float().transpose(-1, -2).contiguous(),
+        output_final_state=True,
+        cu_seqlens=cu_seqlens.to(torch.int32),
```
[the remark this answers is not in the document]
Comment (elwhyjay) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/gated_delta_rule/flashinfer.py]: The high-level chunk_gated_delta_rule converts `cu_seqlens` per path internally (int32 for the sm100 kernel at gdn_prefill.py:320, int64 for sm90 at :348), so forcing int32 here was redundant, not wrong: int32 vs int64 inputs give bit-identical sm100 output (maxdiff 0.0). Removed the forced cast and now pass `cu_seqlens` through so flashinfer selects the per-path dtype.

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/gated_delta_rule.py
Comment (borontion) [inline tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/gated_delta_rule.py:1]: GDN can still be categorized into ops/attention?

Comment (elwhyjay) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/gated_delta_rule.py:1]: Good point. I placed it under `ops/attention/flashinfer` to match the existing `hybrid_linear_attn` / linear-attention structure, where GDN already lives under the attention namespace. Happy to adjust if you prefer another layout.

**Comment by elwhyjay** [comment]
@lightseek-bot Thank you for the reivew! Just to confirm scope: I assume this means removing the  FLA fallback/reference in this PR, not deleting or refactoring the existing mainline `attention/linear` FLA-derived code. If so, unsupported cases currently have no alternate GDN prefill path. Should this PR fail fast there, or keep fallback until the Triton replacement lands? Or If you mean later one please let me know

File: tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/gated_delta_rule.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+_AVAILABLE = False
+
+if current_platform().is_nvidia:
+    try:
+        from flashinfer.gdn_prefill import chunk_gated_delta_rule as _fi_chunk
+
+        _chunk_gated_delta_rule = _fi_chunk
+        _p = current_platform()
+        _sm = _p.arch_version.major * 10 + _p.arch_version.minor
+        _AVAILABLE = _sm == 100
```
[the remark this answers is not in the document]
Comment (elwhyjay) [reply tokenspeed-kernel/python/tokenspeed_kernel/ops/attention/flashinfer/gated_delta_rule.py]: Addressed in d6e8989 alongside the FLA fallback removal - the gate now includes CUDA major and `_has_blackwell_prefill`.

**Comment by elwhyjay** [comment]
@lightseek-bot. Done. Dropped the FLA fallback; unsupported configs (sm100 + CUDA 13 + bf16 + head_dim=128 + head_v == head_k + num_v >= num_q) now reject at the first GDN prefill call.

File: python/tokenspeed/runtime/layers/attention/backends/hybrid_linear_attn.py [hunk 3]
```diff
@@ -1100,17 +1105,34 @@ def forward_extend(
...
-                    g=g,
-                    beta=beta,
+                if not self._gdn_fastpath_checked:
+                    if not (
+                        gdn_flashinfer.is_supported(
+                            head_k_dim, query.dtype, num_heads, num_value_heads
+                        )
+                        and head_v_dim == head_k_dim
+                    ):
+                        raise RuntimeError(
```
[the remark this answers is not in the document]
Comment (elwhyjay) [reply python/tokenspeed/runtime/layers/attention/backends/hybrid_linear_attn.py:1115]: Intentional per maintainer feedback .Unsupported configs now reject instead of falling back.

**Comment by rjzhb** [comment]
Hello! Thanks for sharing this. I was interested in this, so I ran some experiments.

I tested it on 4× B200 with `nvidia/Qwen3.5-397B-A17B-NVFP4`, using the CI perf
config `test/ci/perf/qwen3.5-397b-a17b-nvfp4-evalscope-agentic.yaml` (attn-tp4/
moe-tp4, swe_smith multi-turn dataset via evalscope: `--multi-turn --parallel
1 2 4 8 16 --number 4 8 8 16 32 --max-tokens 500 --extra-args '{"ignore_eos":
true}'`). I disabled MTP for this run so the comparison isolates the GDN prefill
path without speculative accept-rate variance, and just compared the fast path
on vs off on the same server.

Both runs completed the full sweep with 0 failures:

| conc | output tok/s (on → off) | TTFT ms (on → off) | TPOT ms (on → off) |
| ---- | ----------------------- | ------------------ | ------------------ |
| 1    | 162.5 → 161.8           | 466 → 457          | 5.23 → 5.27        |
| 2    | 280.2 → 278.5           | 644 → 663          | 5.53 → 5.53        |
| 4    | 482.0 → 478.7           | 911 → 881          | 6.21 → 6.33        |
| 8    | 704.1 → 688.7           | 1270 → 1253        | 7.59 → 7.88        |
| 16   | 930.7 → 940.8           | 2208 → 2191        | 11.78 → 11.62      |

The differences are within ~±2% and don't point consistently in one direction
(conc 16 is actually slightly slower with the fast path), so at least in this
setup I couldn't see a measurable end-to-end gain. My guess is the workload: it
runs at ~91% KV-cache hit, so only ~2k tokens are actually prefilled per step
(`#new-token 2021, #cached-token 109120`) and the GDN prefill is well under 1% of
wall time. Since the microbench uses uncached 2k–16k prefills, this multi-turn
agentic traffic mostly skips the path being optimized — I'd expect it to help
much more on prefill-heavy / low-cache traffic (long single-turn prompts), which
might be a better case to measure against.

**Comment by minedec** [comment]
Hi, the flashinfer GDN prefill kernel depends on `nvidia-cutlass-dsl-libs-cu13`. Could you add it to tokenspeed-kernel/python/requirements/cuda.txt to make the CI pass?

**Comment by elwhyjay** [comment]
@minedec. Done! Could you rerun the CI? Thanks

**Comment by elwhyjay** [comment]
@minedec thanks! CI fails so I saw the log and it's an ICE compile error in the flashinfer sm100 GDN kernel. I ran into the same thing while working on this PR; it's specific to cutlass-dsl 4.5.2, and 4.5.1 builds fine. Pinned both nvidia-cutlass-dsl and nvidia-cutlass-dsl-libs-cu13 to 4.5.1 and pushed (ce0c93d). Please take a look

FYI the real fix is upstream in flashinfer-ai/flashinfer#3348 (compile-time unroll around the make_tmem_copy loop) - I'll drop the pin once that's merged and released.

**Comment by minedec** [comment]
> @minedec thanks! CI fails so I saw the log and it's an ICE compile error in the flashinfer sm100 GDN kernel. I ran into the same thing while working on this PR; it's specific to cutlass-dsl 4.5.2, and 4.5.1 builds fine. Pinned both nvidia-cutlass-dsl and nvidia-cutlass-dsl-libs-cu13 to 4.5.1 and pushed ([ce0c93d](https://github.com/lightseekorg/tokenspeed/commit/ce0c93dd55c404c0e1f520962fb8951c651452ba)). Please take a look
> 
> FYI the real fix is upstream in [flashinfer-ai/flashinfer#3348](https://github.com/flashinfer-ai/flashinfer/pull/3348) (compile-time unroll around the make_tmem_copy loop) - I'll drop the pin once that's merged and released.

thanks, CI base env has been updated to flashinfer 0.6.11.post3 (verified working with cutlass-dsl 4.5.2), CI has been re-triggered to confirm.
Also, could you add sm103 support to the flashinfer GDN prefill availability check? The B300 CI needs to pass as well.

**Comment by elwhyjay** [comment]
Hi @minedec! Looked at the logs — 3 of the 4 Qwen3.5 jobs still hit the same `tcgen05.make_tmem_copy` ICE, so the upstream bug seems to be still around in this base. The 1 passing job (gb200-4gpu-perf) appears to take a different compile path for some reason. Could you take another look?

**Comment by minedec** [comment]
> Hi @minedec! Looked at the logs — 3 of the 4 Qwen3.5 jobs still hit the same `tcgen05.make_tmem_copy` ICE, so the upstream bug seems to be still around in this base. The 1 passing job (gb200-4gpu-perf) appears to take a different compile path for some reason. Could you take another look?

Sure, I'm looking into the CI environment issue and working on a fix as soon as possible.
Progress is tracked in https://github.com/lightseekorg/tokenspeed/pull/349

**Review by zhyncs — APPROVED** [review]

**Comment by botieking98** [comment]
Is this means Qwen3.5 GDN can not launch on non-Blackwell GPUs ?

**Comment by minedec** [comment]
> Is this means Qwen3.5 GDN can not launch on non-Blackwell GPUs ?

For now, non-Blackwell GPUs do not support Qwen3.5 prefill and will fail fast.
